// @diablosnaps/common/esm/game/utils/item.js
import { Language, } from './../types/index.js';
export function getItemName(id, language, items) {
    return items[language][id];
}
export function getItemTypeLine(itemVariant, itemQuality, itemType, language, translations) {
    let format = '{VARIANT} {QUALITY} {TYPE}';
    switch (language) {
        case Language.German:
        case Language.Polish:
        case Language.Russian:
        case Language.Turkish:
            format = '{VARIANT} {TYPE}';
            break;
        case Language.Mexican:
        case Language.Italian:
        case Language.French:
            format = '{TYPE} {VARIANT}';
            break;
        case Language.Spanish:
        case Language.Portuguese:
            format = '{TYPE} {QUALITY}';
            break;
        case Language.Japanese:
        case Language.SimplifiedChinese:
        case Language.TraditionalChinese:
            format = '{VARIANT}{QUALITY}{TYPE}';
            break;
    }
    return replaceVariables(format, {
        variant: getItemVariantText(itemVariant, language, translations) || '',
        quality: getItemQualityText(itemQuality, language, translations) || '',
        type: getItemTypeText(itemType, language, translations) || '',
    }).trim();
}
export function getItemVariantText(itemVariant, language, translations) {
    return translations[language][`ItemQuality${itemVariant}`];
}
export function getItemQualityText(itemQuality, language, translations) {
    return translations[language][`ItemQuality${itemQuality}`];
}
export function getItemTypeText(itemType, language, translations) {
    return translations[language][`ItemType${itemType}`];
}
export function getItemSocketTypeText(itemSocketType, language, translations) {
    return translations[language][`ItemSocketType${itemSocketType}`];
}
export function getItemAffixName(affixType, affixId, language, affixes) {
    const definition = affixes.definitions[affixType][affixId];
    if (!definition) {
        return '';
    }
    const name = affixes.names[language][definition.name];
    return name;
}
export function getItemAffixDescription(affixType, affixId, language, affixes, value = '#') {
    const definition = affixes.definitions[affixType][affixId];
    if (!definition) {
        return '';
    }
    const description = definition.attributes
        .map(attribute => {
        const attributeDescriptionId = `${attribute.id}${isNaN(attribute.param) ? '' : `#${attribute.param}`}`;
        const attributeDescription = affixes.descriptions[language][attributeDescriptionId];
        return executeFormulas(replaceVariables(attributeDescription, { value }), value);
    })
        .join('\r\n');
    return description;
}
function replaceVariables(template, variables) {
    return template
        .replace(/\{([^}]+)\}/g, (_, key) => {
        const value = variables[key.toLocaleLowerCase()];
        return value === undefined ? `{${key}}` : `${value}`;
    });
}
function roundValue(value, decimals) {
    const factor = Math.pow(10, decimals);
    return Math.round(value * factor) / factor;
}
function formatValue(value, format) {
    switch (format) {
        case '~':
            return `${typeof value === 'string' ? value : Math.round(value)}`;
        case '~%':
            return `${typeof value === 'string' ? value : Math.round(value)}%`;
        case '1':
            return `${typeof value === 'string' ? value : value.toFixed(1)}`;
        case '1%x':
            return `${typeof value === 'string' ? value : value.toFixed(1)}%[x]`;
        case '1%':
        case '%1':
            return `${typeof value === 'string' ? value : value.toFixed(1)}%`;
        case '1%+':
            return `+${typeof value === 'string' ? value : value.toFixed(1)}%`;
        case '2':
            return `${typeof value === 'string' ? value : value.toFixed(2)}`;
        case '2%x':
            return `${typeof value === 'string' ? value : value.toFixed(2)}%[x]`;
        case '2%':
        case '%2':
            return `${typeof value === 'string' ? value : value.toFixed(2)}%`;
        case '2%+':
            return `+${typeof value === 'string' ? value : value.toFixed(2)}%`;
        case '%x':
            return `${typeof value === 'string' ? value : roundValue(value, 3)}%[x]`;
        case '%':
            return `${typeof value === 'string' ? value : roundValue(value, 3)}%`;
        case '+%':
        case '%+':
            return `+${typeof value === 'string' ? value : roundValue(value, 3)}%`;
        case '':
            return `${typeof value === 'string' ? value : roundValue(value, 3)}`;
        default:
            console.warn('Unknown format', format);
            return `${typeof value === 'string' ? value : roundValue(value, 3)}`;
    }
}
function executeFormulas(template, value) {
    return template
        .replace(/\[([^\]]+)\]/g, (_, formula) => {
        if (formula === 'x')
            return '[x]';
        let format = '';
        formula
            .replace(/\}/g, '')
            .replace(/\|([^|]+)\|/g, (_, key) => {
            format = key;
            return '';
        });
        return formatValue(value, format);
    });
}
